napari/layers/surface/surface.py

```diff
@@ -0,0 +1,309 @@
+        """Determine number of dimensions of the layer."""
+        return self.data.shape[1]
+
+    def _get_extent(self):
+        """Determine ranges for slicing given by (min, max, step)."""
+        if len(self.data) == 0:
+            maxs = np.ones(self.data.shape[1], dtype=int)
+            mins = np.zeros(self.data.shape[1], dtype=int)
+        else:
+            maxs = np.max(self.data, axis=0)
+            mins = np.min(self.data, axis=0)
+
+        return [(min, max, 1) for min, max in zip(mins, maxs)]
+
+    @property
+    def colormap(self):
+        """2-tuple of str, vispy.color.Colormap: colormap for luminance images.
+        """
+        return self._colormap_name, self._cmap
+
+    @colormap.setter
+    def colormap(self, colormap):
+        name = '[unnamed colormap]'
+        if isinstance(colormap, str):
+            name = colormap
+        elif isinstance(colormap, tuple):
+            name, cmap = colormap
+            self._colormaps[name] = cmap
+        elif isinstance(colormap, dict):
+            self._colormaps.update(colormap)
+            name = list(colormap)[0]  # first key in dict
+        elif isinstance(colormap, vispy.color.Colormap):
+            name = increment_unnamed_colormap(
+                name, list(self._colormaps.keys())
+            )
+            self._colormaps[name] = colormap
+        else:
+            warnings.warn(f'invalid value for colormap: {colormap}')
+            name = self._colormap_name
+        self._colormap_name = name
+        self._cmap = self._colormaps[name]
+        self._update_thumbnail()
+        self.events.colormap()
+
+    @property
+    def colormaps(self):
+        """tuple of str: names of available colormaps."""
+        return tuple(self._colormaps.keys())
+
+    @property
+    def contrast_limits(self):
+        """list of float: Limits to use for the colormap."""
+        return list(self._contrast_limits)
+
+    @contrast_limits.setter
+    def contrast_limits(self, contrast_limits):
+        self._contrast_limits_msg = (
+            format_float(contrast_limits[0])
+            + ', '
+            + format_float(contrast_limits[1])
+        )
+        self.status = self._contrast_limits_msg
+        self._contrast_limits = contrast_limits
+        if contrast_limits[0] < self._contrast_limits_range[0]:
+            self._contrast_limits_range[0] = copy(contrast_limits[0])
+        if contrast_limits[1] > self._contrast_limits_range[1]:
+            self._contrast_limits_range[1] = copy(contrast_limits[1])
+        self._update_thumbnail()
+        self.events.contrast_limits()
+
+    def _set_view_slice(self):
+        """Sets the view given the indices to slice with."""
+
+        not_disp = list(self.dims.not_displayed)
+        disp = list(self.dims.displayed)
+        indices = np.array(self.dims.indices)
+
+        self._data_view = self.data[:, disp]
+        if len(self.data) == 0:
+            self._view_faces = np.zeros((0, 3))
```

Comment:
This is super weird. I would say we either remove nD support or we need an nD example, because I don't really get how to use this. I think @royerloic was right that we probably need to support different number of vertices for different slices. At any rate, I don't understand how a slider would interact with this stuff?

Reply:
:-) I'm pretty sure the full nD functionality works as expected, but it might not be intuitive / documented / tested properly - i can try improving that, but the way it works right now is to keep the full list of vertices, but trim the list of triangles to only include those triangles that are contained entirely inside the slice.

See this gif where we switch between 2D and 3D display mode where the slider appears:

![nD_surface](https://user-images.githubusercontent.com/6531703/64441800-6b3d8d80-d083-11e9-94fa-d1bdeb970950.gif)

Reply:
Is there a test for this? I suggest:

- create 3D mesh and set ndisplay to 3
- set ndisplay to 2
- iterate over all possible values of the non-displayed dimensions and set the dims to that position

Reply:
I can make better tests

Reply:
i've added a test doing exactly this - inside `test_advanced.py` called `test_changing_display_surface`